### PR TITLE
Incremental JSON <-> struct encoder for implementing preset saving

### PR DESCRIPTION
**Pull request by csboling**
This adds library support for JSON serialization, to be used for saving things like Ansible state to USB disk. For discussion see the lines threads [Preset save to USB disk](https://llllllll.co/t/preset-save-to-usb-disk/10113) and [Serialisation of presets](https://llllllll.co/t/serialisation-of-presets/9466). If this PR is accepted I'll open another to the Ansible repo that uses this encoder to implement the preset save feature.

This code implements an encoder which can convert a JSON document with a known structure to and from a C struct, or do any other kind of processing when recognized pieces of JSON data are found. Tokenizing an incoming JSON document uses a modified version of the [jsmn](https://github.com/zserge/jsmn) library (forked prior to a recent reorganization of that project), which is included. It has the following features which make it suitable for embedded use and easy extensibility.

* it performs no dynamic allocations.
* it can operate incrementally on streaming input, processing tokens which span multiple reads. As such it has almost no minimum text buffer size for storing a portion of text.
* it is able to skip large amounts of whitespace and entire sections of the JSON document that it does not understand, allowing it to load only recognized segments from a much larger document. Supplying all expected properties in an input JSON document is also not required. This allows a module to be fairly forward/backward compatible in the saved preset structures it will accept.
* the processing of the incoming JSON document can be completely customized by supplying different callbacks for handling different nodes. An example data structure used to describe an expected document format can be found in the [tests](https://github.com/csboling/libavr32/blob/93c9ad2e173be0b67a41dbf9070859cf2dc9e4a6/test/unit/json/json_test_schema.c#L10).
* it is agnostic to how it might be used in a larger system, accepting simple [callbacks](https://github.com/csboling/libavr32/blob/93c9ad2e173be0b67a41dbf9070859cf2dc9e4a6/src/json/serdes.h#L57) for reading from an input stream, writing to an output stream, and copying decoded data to its destination. This should make it easy to use the encoder for other possible applications like reading/writing JSON to an I2C channel rather than to a USB disk.

Encoding functions for common data types are included, such as a decimal encoder which handles signed values, and a hex encoder for encoding large binary arrays into JSON as hex strings, allowing relatively easy human editability of the resulting document. Parsing/serializing floating point numbers is not currently implemented. Use of the decimal encoding functions does require static allocation of a 12 byte buffer, I believe this is the only static memory allocation required to use the library.

### CODE REVIEW

**Comment by tehn** [comment]
ambitious and impressive!

> This should make it easy to use the encoder for other possible applications like reading/writing JSON to an I2C channel rather than to a USB disk.

gasp. :)

---

code looks good to me so far. i'll take a second look later today.

File: src/arp.c [hunk 1]
```diff
@@ -132,16 +132,19 @@ void arp_seq_init(arp_seq_t* s) {
 }
 
 bool arp_seq_set_state(arp_seq_t *s, arp_seq_state state) {
+#ifndef RUNNING_TESTS
```
Comment (scanner-darkly) [inline src/arp.c]: curious, why is this needed? does it need to be done in other places that use `irqs_pause/resume`? if so, would it be better to move `ifndef` to the actual functions?

**Comment by scanner-darkly** [comment]
this is great - super useful to have it be part of libavr32.

File: src/arp.c (hunk 1 above)
Comment (csboling) [reply src/arp.c]: I just found [this](https://github.com/monome/libavr32/pull/44#issuecomment-458303142) conversation which leads me to believe that this `#ifndef` isn't needed on the latest master. This was how I made the tests pass when I first made this branch, I can clean this up.

Comment (scanner-darkly) [reply src/arp.c]: ah right, this rings a bell. i think either way is fine - `irqs` stuff is not needed for running tests anyway.

Comment (csboling) [reply src/arp.c]: Removed `RUNNING_TESTS`, realized there was actually an error testing `json_read_object_cached`, added a make target so that if you have valgrind installed you can run `make leaks`, and found some leaks in the test code - nothing actually affecting the JSON encoder itself I think.
